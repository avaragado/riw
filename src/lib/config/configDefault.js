// @flow

const cfg: RIWConfig = {
    configFile: null, // overridden
    rootDir: '', // overridden
    defaultLocale: 'en-us',
    targetLocales: [],
    translationsDatabaseFile: 'src/locale/riw-db.json',
    sourceDirs: ['src/**/*.js'],
    collateDir: 'tmp/babel-plugin-react-intl',
    inputMode: 'source',
    translationsOutputFile: 'src/locale/[locale].json',
    outputMode: 'file-per-locale',
    todoFile: 'src/locale/TODO-untranslated.json',
};

export default cfg;
