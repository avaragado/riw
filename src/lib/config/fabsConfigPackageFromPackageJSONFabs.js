// @flow

import path from 'path';

export default (fabsPackageJSON: ?AbsolutePath): ?AbsolutePath => (
    fabsPackageJSON
        ? path.resolve(path.dirname(fabsPackageJSON), '.riw-config.js')
        : null
    );
