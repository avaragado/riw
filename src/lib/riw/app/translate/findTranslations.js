// @flow

import chain from 'ramda/src/chain';
import compose from 'ramda/src/compose';
import map from 'ramda/src/map';
import reduce from 'ramda/src/reduce';
import pathOr from 'ramda/src/pathOr';
import assocPath from 'ramda/src/assocPath';

import type {
    LocaleId,
    MessageId,
    TranslatedMessage,
    MessageDescriptorWithFile,
    UntranslatedMessageDescriptor,
} from '../../../../types';
import type { Config } from '../../../config';
import type { Notifier } from '../../../notify';
import type { TranslationsDB } from '../../db';
import { sDescriptionDefault } from '../../';

type LocaleMapper = (config: Config) => (md: MessageDescriptorWithFile) =>
    UntranslatedMessageDescriptor;

type Lookup = {|
    mdu: UntranslatedMessageDescriptor,
    result: TranslatedMessage | null,
|};

export type TranslationLookupResult = {|
    locale: {
        [key: LocaleId]: {
            [key: MessageId]: TranslatedMessage,
        },
    },
    todos: UntranslatedMessageDescriptor[],
|};

type LookerUpper = (config: Config, db: TranslationsDB) =>
    (mdu: UntranslatedMessageDescriptor) => Lookup;

type LookupReducer = (acc: TranslationLookupResult, lookup: Lookup) =>
    TranslationLookupResult;

type TranslationFinder = (
    config: Config,
    notify: Notifier,
    db: TranslationsDB
) => (armd: MessageDescriptorWithFile[]) => TranslationLookupResult;


const resultEmpty: TranslationLookupResult = {
    locale: {},
    todos: [],
};

const eachLocale: LocaleMapper = config => md => map(
    locale => ({ ...md, locale }),
    config.targetLocales.concat(config.defaultLocale),
);

const includeLookup: LookerUpper = (config, db) => mdu => ({
    mdu,
    result: mdu.locale === config.defaultLocale
        ? mdu.defaultMessage
        : pathOr(null, [
            'data',
            mdu.defaultMessage,
            mdu.description || sDescriptionDefault,
            mdu.locale,
        ], db),
});

const reduceLookup: LookupReducer = (acc, lookup) => {
    const { mdu, result } = lookup;

    if (result) {
        return assocPath(['locale', mdu.locale, mdu.id], result, acc);
    }

    return ({
        locale: acc.locale,
        todos: acc.todos.concat(mdu),
    }: TranslationLookupResult);
};

const find: TranslationFinder = (config, notify, db) => compose(
    notify('endLookup'),
    reduce(reduceLookup, resultEmpty),
    map(includeLookup(config, db)),
    chain(eachLocale(config)),
    notify('startLookup'),
);

export default find;
