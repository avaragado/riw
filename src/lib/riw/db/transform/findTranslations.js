// @flow

import filter from 'ramda/src/filter';

import makeQuadMatcher from './util/makeQuadMatcher';

// eslint-disable-next-line no-unused-vars
const transformer: RIWDBQuadsTransformer = (config, opt?: RIWCLIOptDBFind) => quads =>
    (opt
        ? filter(makeQuadMatcher(opt.match), quads)
        : quads
    );

export default transformer;
