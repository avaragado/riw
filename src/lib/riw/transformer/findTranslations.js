// @flow

const makeFilter = match => (quad: RIWDBQuad): boolean => (
    (!match.defaultMessage || quad[0] === match.defaultMessage) &&
    (!match.description || quad[1] === match.description) &&
    (!match.locale || quad[2] === match.locale) &&
    (!match.translation || quad[3] === match.translation)
);

// eslint-disable-next-line no-unused-vars
const transformer: RIWDBQuadsTransformer = (config, opt?: RIWCLIOptDBFind) => quads =>
    (opt
        ? quads.filter(makeFilter(opt.match))
        : quads
    );

export default transformer;
