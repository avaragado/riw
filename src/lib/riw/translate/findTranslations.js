// @flow

import chain from 'ramda/src/chain';
import compose from 'ramda/src/compose';
import map from 'ramda/src/map';
import reduce from 'ramda/src/reduce';
import pathOr from 'ramda/src/pathOr';
import assocPath from 'ramda/src/assocPath';

import { sDescriptionDefault } from '../';

type LocaleMapper = (config: RIWConfig) => (md: RIWMessageDescriptor) =>
    RIWMessageDescriptorUntranslated;

type TranslationLookup = {
    mdu: RIWMessageDescriptorUntranslated,
    result: RIWDBTranslatedMessage | null,
};

type LookerUpper = (config: RIWConfig, db: RIWDB) => (mdu: RIWMessageDescriptorUntranslated) =>
    TranslationLookup;

type LookupReducer = (acc: RIWFindTranslationResult, lookup: TranslationLookup) =>
    RIWFindTranslationResult;

type TranslationFinder = <T>(
    config: RIWConfig,
    notify: string => T => T,
    db: RIWDB
) => (armd: RIWMessageDescriptor[]) => RIWFindTranslationResult;


const resultEmpty: RIWFindTranslationResult = {
    locale: {},
    armdu: [],
};

const eachLocale: LocaleMapper = config => md => map(
    locale => ({ ...md, locale }),
    config.targetLocales.concat(config.defaultLocale),
);

const includeLookup: LookerUpper = (config, db) => mdu => ({
    mdu,
    result: mdu.locale === config.defaultLocale
        ? mdu.defaultMessage
        : pathOr(null, [
            'data',
            mdu.defaultMessage,
            mdu.description || sDescriptionDefault,
            mdu.locale,
        ], db),
});

const reduceLookup: LookupReducer = (acc, lookup) => {
    const { mdu, result } = lookup;

    if (result) {
        return assocPath(['locale', mdu.locale, mdu.id], result, acc);
    }

    return {
        ...acc,
        armdu: acc.armdu.concat(mdu),
    };
};

const find: TranslationFinder = (config, notify, db) => compose(
    notify('endLookup'),
    reduce(reduceLookup, resultEmpty),
    map(includeLookup(config, db)),
    chain(eachLocale(config)),
    notify('startLookup'),
);

export default find;
