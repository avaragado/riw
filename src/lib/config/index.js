// @flow

import configDefault from './resolve/configDefault';
import fabsConfigPackageForCWD from './fromPackage/fabsConfigPackageForCWD';
import configFromPath from './fromPath';
import configFromPackage from './fromPackage';
import configResolve from './resolve';

export {
    configDefault,
    fabsConfigPackageForCWD,
    configFromPath,
    configFromPackage,
    configResolve,
};
