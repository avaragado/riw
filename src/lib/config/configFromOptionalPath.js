// @flow

import path from 'path';

import log from '../log';

import configDefault from './configDefault';
import fabsPackageJSONFromDabs from './fabsPackageJSONFromDabs';
import fabsConfigPackageFromPackageJSONFabs from './fabsConfigPackageFromPackageJSONFabs';

const requireOrNull = (fabs: AbsolutePath): ?Object => {
    try {
        return require(fabs);

    } catch (err) {
        return null;
    }
};

const configFromPath = (pathConfig: Path): ?RIWConfig => {
    const fabs: AbsolutePath = path.resolve(pathConfig);
    const configOverride = requireOrNull(fabs);

    if (configOverride === null) {
        log.error('riw', 'File not found: %j.', fabs);
        return null;
    }

    return {
        ...configDefault,
        ...configOverride,
        rootDir: path.dirname(fabs),
    };
};

const configFromPackage = (): ?RIWConfig => {
    const fabsPackageJSON = fabsPackageJSONFromDabs(process.cwd());
    const fabsConfigPackage = fabsConfigPackageFromPackageJSONFabs(fabsPackageJSON);

    if (!fabsConfigPackage || !fabsPackageJSON) {
        log.error('riw', 'Unable to find package root');
        return null;
    }

    let configOverride: ?Object = requireOrNull(fabsConfigPackage);

    if (configOverride === null) {
        configOverride = requireOrNull(fabsPackageJSON);

        if (configOverride && configOverride.riw) {
            configOverride = configOverride.riw;
        } else {
            configOverride = null;
        }
    }

    if (configOverride === null) {
        log.warn('riw', 'No config located for the project. You probably want to add some configuration.');
        configOverride = {};
    }

    return {
        ...configDefault,
        ...configOverride,
        rootDir: path.dirname(fabsConfigPackage),
    };

};

export default (pathConfig: ?Path): ?RIWConfig => (
    pathConfig
        ? configFromPath(pathConfig)
        : configFromPackage()
);
