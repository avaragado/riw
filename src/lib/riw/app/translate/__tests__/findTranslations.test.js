// @flow

import type { MessageDescriptorWithFile } from '../../../../../types';
import type { ConfigSparseWithSource } from '../../../../config';
import { configResolve } from '../../../../config';
import type { TranslationsDB } from '../../../db';

import findTranslations from '../findTranslations';

const notify = () => x => x;

type Fixture = {
    name: string,
    db: TranslationsDB,
    armd: MessageDescriptorWithFile[],
    configOverride: ConfigSparseWithSource,
};

const fixtures: Fixture[] = [
    {
        name: '01 nothing gets nothing',
        db: {
            version: 1,
            data: {
            },
        },
        armd: [],
        configOverride: {
            defaultLocale: 'en-aa',
        },
    },

    {
        name: '02 no db: md misses',
        db: {
            version: 1,
            data: {
            },
        },
        armd: [
            {
                id: 'md1',
                defaultMessage: 'a',
                file: '/a.js',
            },
        ],
        configOverride: {
            defaultLocale: 'en-aa',
            targetLocales: ['en-up'],
        },
    },

    {
        name: '03 db missing target locale: md ignored',
        db: {
            version: 1,
            data: {
                'a': {
                    '_': {
                        'en-up': 'A',
                    },
                },
            },
        },
        armd: [
            {
                id: 'md1',
                defaultMessage: 'a',
                file: '/a.js',
            },
        ],
        configOverride: {
            defaultLocale: 'en-aa',
        },
    },

    {
        name: '04 db has missing defaultMessage: md misses',
        db: {
            version: 1,
            data: {
                'NOT': {
                    'a desc': {
                        'en-up': 'A',
                    },
                },
            },
        },
        armd: [
            {
                id: 'md1',
                defaultMessage: 'a',
                description: 'a desc',
                file: '/a.js',
            },
        ],
        configOverride: {
            defaultLocale: 'en-aa',
            targetLocales: ['en-up'],
        },
    },

    {
        name: '05 db has missing defined string description: md misses',
        db: {
            version: 1,
            data: {
                'a': {
                    '_': {
                        'en-up': 'A',
                    },
                    'NOT': {
                        'en-up': 'A NOT',
                    },
                },
            },
        },
        armd: [
            {
                id: 'md1',
                defaultMessage: 'a',
                description: 'a desc',
                file: '/a.js',
            },
        ],
        configOverride: {
            defaultLocale: 'en-aa',
            targetLocales: ['en-up'],
        },
    },

    {
        name: '06 db has missing defined object description: md misses',
        db: {
            version: 1,
            data: {
                'a': {
                    '_': {
                        'en-up': 'A',
                    },
                    'NOT': {
                        'en-up': 'A NOT',
                    },
                },
            },
        },
        armd: [
            {
                id: 'md1',
                defaultMessage: 'a',
                description: {
                    blah: 'foo',
                },
                file: '/a.js',
            },
        ],
        configOverride: {
            defaultLocale: 'en-aa',
            targetLocales: ['en-up'],
        },
    },

    {
        name: '07 db has missing locale: md misses',
        db: {
            version: 1,
            data: {
                'a': {
                    'a desc': {
                        'en-zz': 'zz A',
                    },
                },
            },
        },
        armd: [
            {
                id: 'md1',
                defaultMessage: 'a',
                description: 'a desc',
                file: '/a.js',
            },
        ],
        configOverride: {
            defaultLocale: 'en-aa',
            targetLocales: ['en-up'],
        },
    },

    {
        name: '08 match works - string description',
        db: {
            version: 1,
            data: {
                'a': {
                    '_': {
                        'en-up': 'NOT A',
                    },
                    'a desc': {
                        'en-up': 'A',
                    },
                    'other desc': {
                        'en-up': 'NOT A',
                    },
                },
            },
        },
        armd: [
            {
                id: 'md1',
                defaultMessage: 'a',
                description: 'a desc',
                file: '/a.js',
            },
        ],
        configOverride: {
            defaultLocale: 'en-aa',
            targetLocales: ['en-up'],
        },
    },

    {
        name: '09 match works - object description',
        db: {
            version: 1,
            data: {
                'a': {
                    '{"blah":"foo"}': {
                        'en-up': 'A',
                    },
                    'blah': {
                        'en-up': 'NOT A',
                    },
                    'foo': {
                        'en-up': 'NOT A',
                    },
                },
            },
        },
        armd: [
            {
                id: 'md1',
                defaultMessage: 'a',
                description: {
                    blah: 'foo',
                },
                file: '/a.js',
            },
        ],
        configOverride: {
            defaultLocale: 'en-aa',
            targetLocales: ['en-up'],
        },
    },

    {
        name: '10 match works - no description',
        db: {
            version: 1,
            data: {
                'a': {
                    '_': {
                        'en-up': 'A',
                    },
                    'a desc': {
                        'en-up': 'NOT A',
                    },
                    'other desc': {
                        'en-up': 'NOT A',
                    },
                },
            },
        },
        armd: [
            {
                id: 'md1',
                defaultMessage: 'a',
                file: '/a.js',
            },
        ],
        configOverride: {
            defaultLocale: 'en-aa',
            targetLocales: ['en-up'],
        },
    },

    {
        name: '11 match works - undefined description',
        db: {
            version: 1,
            data: {
                'a': {
                    '_': {
                        'en-up': 'A',
                    },
                    'a desc': {
                        'en-up': 'NOT A',
                    },
                    'other desc': {
                        'en-up': 'NOT A',
                    },
                },
            },
        },
        armd: [
            {
                id: 'md1',
                defaultMessage: 'a',
                description: undefined,
                file: '/a.js',
            },
        ],
        configOverride: {
            defaultLocale: 'en-aa',
            targetLocales: ['en-up'],
        },
    },

    {
        name: '12 simple multi',
        db: {
            version: 1,
            data: {
                'a1': {
                    '_': {
                        'en-up': 'A1',
                        'en-ne': '1A',
                    },
                    'a1 desc': {
                        'en-up': 'A1 DESC',
                    },
                },
                'b1': {
                    '_': {
                        'en-up': 'B1',
                        'en-ne': '1B',
                    },
                },
                'c1': {
                    '_': {
                        'en-up': 'C1',
                        'en-zz': 'zz C1',
                    },
                    'NOT': {
                        'en-up': 'C1 NOT',
                    },
                },
            },
        },
        armd: [
            {
                id: 'md1',
                defaultMessage: 'a1',
                file: '/a.js',
            },
            {
                id: 'md2',
                defaultMessage: 'c1',
                file: '/c.js',
            },
            {
                id: 'md3',
                defaultMessage: 'a1',
                description: 'a1 desc',
                file: '/a.js',
            },
        ],
        configOverride: {
            defaultLocale: 'en-aa',
            targetLocales: ['en-up', 'en-ne'],
        },
    },
];

describe('lib/riw/app/translate/findTranslations', () => {
    fixtures.forEach((fixture) => {
        it(fixture.name, () => {
            const cfg = configResolve(fixture.configOverride);

            const received = findTranslations(cfg, notify, fixture.db)(fixture.armd);

            expect(received).toMatchSnapshot();
        });
    });
});
