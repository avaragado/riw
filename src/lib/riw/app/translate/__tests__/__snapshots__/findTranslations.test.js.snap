// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lib/riw/app/translate/findTranslations 01 nothing gets nothing 1`] = `
Object {
  "locale": Object {},
  "todos": Array [],
}
`;

exports[`lib/riw/app/translate/findTranslations 02 no db: md misses 1`] = `
Object {
  "locale": Object {
    "en-aa": Object {
      "md1": "a",
    },
  },
  "todos": Array [
    Object {
      "defaultMessage": "a",
      "file": "/a.js",
      "id": "md1",
      "locale": "en-up",
    },
  ],
}
`;

exports[`lib/riw/app/translate/findTranslations 03 db missing target locale: md ignored 1`] = `
Object {
  "locale": Object {
    "en-aa": Object {
      "md1": "a",
    },
  },
  "todos": Array [],
}
`;

exports[`lib/riw/app/translate/findTranslations 04 db has missing defaultMessage: md misses 1`] = `
Object {
  "locale": Object {
    "en-aa": Object {
      "md1": "a",
    },
  },
  "todos": Array [
    Object {
      "defaultMessage": "a",
      "description": "a desc",
      "file": "/a.js",
      "id": "md1",
      "locale": "en-up",
    },
  ],
}
`;

exports[`lib/riw/app/translate/findTranslations 05 db has missing defined string description: md misses 1`] = `
Object {
  "locale": Object {
    "en-aa": Object {
      "md1": "a",
    },
  },
  "todos": Array [
    Object {
      "defaultMessage": "a",
      "description": "a desc",
      "file": "/a.js",
      "id": "md1",
      "locale": "en-up",
    },
  ],
}
`;

exports[`lib/riw/app/translate/findTranslations 06 db has missing defined object description: md misses 1`] = `
Object {
  "locale": Object {
    "en-aa": Object {
      "md1": "a",
    },
  },
  "todos": Array [
    Object {
      "defaultMessage": "a",
      "description": Object {
        "blah": "foo",
      },
      "file": "/a.js",
      "id": "md1",
      "locale": "en-up",
    },
  ],
}
`;

exports[`lib/riw/app/translate/findTranslations 07 db has missing locale: md misses 1`] = `
Object {
  "locale": Object {
    "en-aa": Object {
      "md1": "a",
    },
  },
  "todos": Array [
    Object {
      "defaultMessage": "a",
      "description": "a desc",
      "file": "/a.js",
      "id": "md1",
      "locale": "en-up",
    },
  ],
}
`;

exports[`lib/riw/app/translate/findTranslations 08 match works - string description 1`] = `
Object {
  "locale": Object {
    "en-aa": Object {
      "md1": "a",
    },
    "en-up": Object {
      "md1": "A",
    },
  },
  "todos": Array [],
}
`;

exports[`lib/riw/app/translate/findTranslations 09 match works - object description 1`] = `
Object {
  "locale": Object {
    "en-aa": Object {
      "md1": "a",
    },
    "en-up": Object {
      "md1": "A",
    },
  },
  "todos": Array [],
}
`;

exports[`lib/riw/app/translate/findTranslations 10 match works - no description 1`] = `
Object {
  "locale": Object {
    "en-aa": Object {
      "md1": "a",
    },
    "en-up": Object {
      "md1": "A",
    },
  },
  "todos": Array [],
}
`;

exports[`lib/riw/app/translate/findTranslations 11 match works - undefined description 1`] = `
Object {
  "locale": Object {
    "en-aa": Object {
      "md1": "a",
    },
    "en-up": Object {
      "md1": "A",
    },
  },
  "todos": Array [],
}
`;

exports[`lib/riw/app/translate/findTranslations 12 simple multi 1`] = `
Object {
  "locale": Object {
    "en-aa": Object {
      "md1": "a1",
      "md2": "c1",
      "md3": "a1",
    },
    "en-ne": Object {
      "md1": "1A",
    },
    "en-up": Object {
      "md1": "A1",
      "md2": "C1",
      "md3": "A1 DESC",
    },
  },
  "todos": Array [
    Object {
      "defaultMessage": "c1",
      "file": "/c.js",
      "id": "md2",
      "locale": "en-ne",
    },
    Object {
      "defaultMessage": "a1",
      "description": "a1 desc",
      "file": "/a.js",
      "id": "md3",
      "locale": "en-ne",
    },
  ],
}
`;
